llvm/lib/IR/Instruction.cpp

```diff
@@ -218,7 +218,7 @@ void Instruction::moveBeforeImpl(BasicBlock &BB, InstListType::iterator I,
 
     // If we're inserting at point I, and not in front of the DPValues attached
     // there, then we should absorb the DPValues attached to I.
-    if (!InsertAtHead)
+    if (NextMarker && !InsertAtHead)
```

Comment:
Perhaps this check `NextMarker != nullptr` should move outside this if, to prevent creating a marker on `this` if there's nothing to absorb? YMMV, not essential.